kaldb/src/main/java/com/slack/kaldb/chunk/manager/ChunkManager.java

```diff
@@ -0,0 +1,166 @@
+import com.slack.kaldb.logstore.search.SearchResult;
+import com.slack.kaldb.logstore.search.SearchResultAggregator;
+import com.slack.kaldb.logstore.search.SearchResultAggregatorImpl;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.spotify.futures.CompletableFutures;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ChunkManager<T> extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(ChunkManager.class);
+
+  protected final Map<String, Chunk<T>> chunkMap = new ConcurrentHashMap<>();
+
+  // TODO: We want to move this to the config eventually
+  // Less than KaldbDistributedQueryService#READ_TIMEOUT_MS
+  public static final int QUERY_TIMEOUT_SECONDS = 10;
+  public static final int LOCAL_QUERY_THREAD_POOL_SIZE = 4;
+
+  private static final ExecutorService queryExecutorService = queryThreadPool();
+
+  /*
+     One day we will have to think about rate limiting/backpressure and we will revisit this so it could potentially reject threads if the pool is full
+  */
+  private static ExecutorService queryThreadPool() {
+    return Executors.newFixedThreadPool(
+        LOCAL_QUERY_THREAD_POOL_SIZE,
+        new ThreadFactoryBuilder().setNameFormat("chunk-manager-query-%d").build());
+  }
+
+  /*
+   * Query the chunks in the time range, aggregate the results per aggregation policy and return the results.
+   * We aggregate locally and and then the query aggregator will aggregate again. This is OKAY for the current use-case we support
+   * 1. topK results sorted by timestamp
+   * 2. histogram over a fixed time range
+   * We will not aggregate locally for future use-cases that have complex group by etc
+   */
+  public CompletableFuture<SearchResult<T>> query(SearchQuery query) {
+
+    SearchResult<T> errorResult =
+        new SearchResult<>(new ArrayList<>(), 0, 0, new ArrayList<>(), 0, 0, 1, 0);
+
+    List<CompletableFuture<SearchResult<T>>> queries =
+        chunkMap
+            .values()
+            .stream()
+            .filter(
+                chunk ->
+                    chunk.containsDataInTimeRange(
+                        query.startTimeEpochMs / 1000, query.endTimeEpochMs / 1000))
+            .map(
+                (chunk) ->
+                    CompletableFuture.supplyAsync(
+                            () -> {
+                              ScopedSpan span =
+                                  Tracing.currentTracer()
+                                      .startScopedSpan("ReadWriteChunkImpl.query");
+                              span.tag("chunkId", chunk.info().chunkId);
```

Comment:
I don't think it is valuable to add all the chunk metadata to the tag. I think just adding the chunkId is fine. We can look up metadata store if needed for additional info and add tags as needed. For now, let's drop the rest of the tags. 

Is the query encoded in the trace any where?

Reply:
These were added as part of #91, and were exceptionally useful when troubleshooting the performance issue we had with the Armeria threadpools.

You can see various screenshots on that PR of the tags we added to the traces. The chunkID is helpful when troubleshooting as that directly maps to the blob store directory.